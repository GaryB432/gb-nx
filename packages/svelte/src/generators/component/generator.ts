import {
  formatFiles,
  generateFiles,
  getProjects,
  names,
  readNxJson,
  updateNxJson,
  type Tree,
} from '@nx/devkit';
import { join } from 'path';
import {
  getSvelteConfig,
  getSvelteFiles,
  supportsRunes,
} from '../../utils/svelte';
import type { Schema as ComponentGeneratorSchema } from './schema';

export interface NormalizedComponentSchema extends ComponentGeneratorSchema {
  projectLibRoot: string;
  projectName: string;
}

function normalizeOptions(
  tree: Tree,
  options: ComponentGeneratorSchema
): NormalizedComponentSchema {
  const ws = readNxJson(tree);
  const projectName = options.project ?? ws?.defaultProject;
  const projects = getProjects(tree);
  if (!projectName) {
    throw new Error('no project');
  }
  const project = projects.get(projectName);

  if (!project) {
    throw new Error(`Project "${projectName}" not found`);
  }

  const runesSupport = supportsRunes(tree, project);

  if (options.runes && !runesSupport.supports) {
    throw new Error(
      `runes feature requires svelte >= 5 (currently '${runesSupport.svelte}')`
    );
  }

  const runes = options.runes ?? runesSupport.supports;

  const config = getSvelteConfig(tree, project);

  if (!config) {
    throw new Error(`Project "${projectName}" not configured for svelte`);
  }

  // TODO choose lib or routes (new sverdle has components next to route files)
  const { lib } = getSvelteFiles(config);
  options.language = options.language ?? 'js';
  options.style = options.style ?? 'css';
  return {
    ...options,
    runes,
    projectName,
    projectLibRoot: join(project.root, lib),
  };
}

function addFiles(tree: Tree, options: NormalizedComponentSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    tmpl: '',
  };
  generateFiles(
    tree,
    join(__dirname, './files'),
    join(options.projectLibRoot, options.directory ?? ''),
    templateOptions
  );
}

function addGeneratorDefaults(
  tree: Tree,
  options: NormalizedComponentSchema
): void {
  const nxJson = readNxJson(tree);

  if (!nxJson) {
    return;
  }

  const { directory, language, style } = options;
  nxJson.generators = nxJson.generators ?? {};
  nxJson.generators['@gb-nx/svelte:component'] = {
    directory,
    language,
    style,
    ...(nxJson.generators['@gb-nx/svelte:component'] || {}),
  };

  updateNxJson(tree, nxJson);
}

export default async function componentGenerator(
  tree: Tree,
  options: ComponentGeneratorSchema
): Promise<void> {
  const normalizedOptions = normalizeOptions(tree, options);
  addFiles(tree, normalizedOptions);
  addGeneratorDefaults(tree, normalizedOptions);
  if (!normalizedOptions.skipFormat) {
    await formatFiles(tree);
  }
}
