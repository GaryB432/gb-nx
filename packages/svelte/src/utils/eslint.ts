import {
  joinPathFragments,
  type ProjectConfiguration,
  type Tree,
} from '@nx/devkit';

export interface EslintConfiguration {
  env?: {
    browser: boolean;
    es2021: boolean;
    node: boolean;
  };
  extends?: string[];
  ignorePatterns?: string[];
  overrides: unknown[];
  parser?: string;
  parserOptions?: {
    ecmaVersion: string;
    extraFileExtensions: string[];
    project: string;
    sourceType: string;
  };
  plugins: string[];
  root: boolean;
  rules?: unknown;
}

function updateWorkspaceEslint(tree: Tree): void {
  const config: Partial<EslintConfiguration> = {
    root: true,
    env: {
      browser: true,
      es2021: true,
      node: true,
    },
    extends: ['plugin:gb/recommended', 'plugin:svelte/recommended'],
    overrides: [
      {
        files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
        rules: {
          '@nx/enforce-module-boundaries': [
            'error',
            {
              enforceBuildableLibDependency: true,
              allow: [],
              depConstraints: [
                {
                  sourceTag: '*',
                  onlyDependOnLibsWithTags: ['*'],
                },
              ],
            },
          ],
        },
      },
      {
        files: ['*.svelte'],
        parser: 'svelte-eslint-parser',
        parserOptions: {
          parser: '@typescript-eslint/parser',
        },
      },
    ],
    parser: '@typescript-eslint/parser',
    parserOptions: {
      ecmaVersion: 'latest',
      sourceType: 'module',
      project: 'tsconfig.base.json',
      extraFileExtensions: ['.svelte'],
    },
    plugins: ['@nx', 'gb'],
    rules: {},
  };
  tree.write('.eslintrc.json', JSON.stringify(config));
}
function updateProjectEslint(
  tree: Tree,
  project: Pick<ProjectConfiguration, 'root'>
): void {
  const config: Partial<EslintConfiguration> = {
    extends: ['../../.eslintrc.json'],
    ignorePatterns: ['.svelte-kit', 'build/*', '**/*.config.*'],
  };

  tree.write(
    joinPathFragments(project.root, '.eslintrc.json'),
    JSON.stringify(config)
  );
}

/**
 * 
 * @param tree 
 * @param project 
 * @deprecated
 */
export function updateEslint(
  tree: Tree,
  project: Pick<ProjectConfiguration, 'root'>
): void {
  updateWorkspaceEslint(tree);
  updateProjectEslint(tree, project);
}
