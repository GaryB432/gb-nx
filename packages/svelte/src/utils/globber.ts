import { parse } from 'path';

export function includes(path: string, globs: string[]): boolean {
  const parsedPath = parse(path);
  for (const c of globs) {
    if (c === path) {
      return true;
    }
    const glob = parse(c);
    if (
      glob.base === '*' &&
      glob.root === parsedPath.root &&
      glob.dir === parsedPath.dir
    ) {
      return true;
    }
  }
  return false;
}
