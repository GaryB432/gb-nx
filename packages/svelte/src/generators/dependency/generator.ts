import type { ProjectConfiguration, Tree } from '@nrwl/devkit';
import {
  formatFiles,
  getProjects,
  joinPathFragments,
  offsetFromRoot,
  // output,
  readWorkspaceConfiguration,
} from '@nrwl/devkit';

import { makeAliasName, type NamedPath } from '../../utils/paths';
import { getSvelteConfig } from '../../utils/svelte';
import { addToSvelteConfiguration, getConfiguredAliases } from './alias';
import type { Schema as DependencyGeneratorSchema } from './schema';

function updateSvelteConfig(
  tree: Tree,
  project: ProjectConfiguration,
  alias: NamedPath
): void {
  const fname = joinPathFragments(project.root, 'svelte.config.js');
  const config = tree.read(fname);
  if (!config) {
    return;
  }
  tree.write(fname, addToSvelteConfiguration(config.toString(), alias));
}

export default async function (
  tree: Tree,
  schema: DependencyGeneratorSchema
): Promise<void> {
  const notfound = (p: string) => `project '${p}' was not found in workspace`;
  const notSvelte = (p: string) =>
    `project '${p}' is not configured for svelte`;
  const ws = readWorkspaceConfiguration(tree);
  const projName = schema.project ?? ws.defaultProject;
  const aliasScope = schema.scope ?? ws.npmScope;

  const projects = getProjects(tree);
  const project = projects.get(projName);

  if (!project) {
    throw new Error(notfound(projName));
  }

  const dep = projects.get(schema.dependency);
  if (!dep) {
    throw new Error(notfound(schema.dependency));
  }

  const psrc = project.sourceRoot;
  const dsrc = dep.sourceRoot;

  const config = getSvelteConfig(tree, project);

  if (!config) {
    throw new Error(notSvelte(projName));
  }

  const aliases = getConfiguredAliases(config).slice(0);
  const depName = makeAliasName(schema.dependency, aliasScope);

  if (dsrc && psrc) {
    const path = joinPathFragments(offsetFromRoot(dep.root), dsrc);

    const aliasToAdd = { name: depName, path };
    aliases.push(aliasToAdd);

    // output.log({
    //   title: 'Configured Aliases',
    //   bodyLines: aliases.map(
    //     (p) =>
    //       `${output.colors.white(p.name)}: "${output.colors.green(p.path)}"`
    //   ),
    // });

    updateSvelteConfig(tree, project, aliasToAdd);
  }

  const pjbuff = tree.read(joinPathFragments(project.root, 'package.json'));
  if (pjbuff) {
    const pkg = JSON.parse(pjbuff.toString()) as {
      name: string;
      nx: { implicitDependencies: string[] };
    };
    const ideps = pkg.nx.implicitDependencies ?? [];
    if (!ideps.includes(schema.dependency)) {
      pkg.nx.implicitDependencies = [...ideps, schema.dependency];
      tree.write(
        joinPathFragments(project.root, 'package.json'),
        JSON.stringify(pkg, undefined, 2).concat('\n')
      );
    }
  }
  await formatFiles(tree);
}
