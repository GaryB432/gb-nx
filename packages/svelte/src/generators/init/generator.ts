import {
  addDependenciesToPackageJson,
  formatFiles,
  runTasksInSerial,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';
import { prettierPluginSvelteVersion } from '../../utils/versions';
import type { Schema } from './schema';

function updateDependencies(tree: Tree) {
  return addDependenciesToPackageJson(
    tree,
    {},
    {
      'prettier-plugin-svelte': prettierPluginSvelteVersion,
    }
  );
}

export default async function initGenerator(
  host: Tree,
  schema: Schema
): Promise<GeneratorCallback> {
  const tasks: GeneratorCallback[] = [];
  const installTask = updateDependencies(host);
  tasks.push(installTask);

  if (!schema.skipFormat) {
    await formatFiles(host);
  }
  return runTasksInSerial(...tasks);
}
