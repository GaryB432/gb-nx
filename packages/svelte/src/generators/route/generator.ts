import {
  formatFiles,
  getProjects,
  joinPathFragments,
  names,
  readNxJson,
  type ProjectConfiguration,
  type Tree,
} from '@nx/devkit';
import { getSvelteConfig, getSvelteFiles } from '../../utils/svelte';
import type { Schema } from './schema';

interface NormalizedSchema extends Schema {
  routePath: string;
}

function getRouteParam(typedParam: string): [string, string] {
  const r = typedParam.split('=', 2);
  return [r[0], r[1]];
}

function normalizeOptions(
  host: Tree,
  options: Schema
): Required<NormalizedSchema> {
  const ws = readNxJson(host);
  options.project = options.project ?? ws?.defaultProject;
  options.directory = options.directory ?? '';

  if (!options.project) {
    throw new Error('Project or defaultProject required');
  }

  const defaultOptions: Required<NormalizedSchema> = {
    directory: options.directory,
    language: 'js',
    load: 'none',
    name: options.name,
    project: options.project,
    skipFormat: false,
    skipTests: false,
    style: 'css',
    routePath: options.name,
  };
  return {
    ...defaultOptions,

    ...options,
    routePath: joinPathFragments(options.directory ?? '', options.name),
  };
}

interface Segment {
  isParam: boolean;
  paramType: string | undefined;
  path: string;
}

export function getSegments(
  options: Pick<NormalizedSchema, 'routePath'>
): Segment[] {
  return options.routePath.split('/').map<Segment>((part) => {
    const isParam = part.startsWith('[');
    const [path, paramType] = getRouteParam(
      isParam ? part.slice(1, part.length - 1) : part
    );

    return { isParam, path, paramType };
  });
}
function addLoadPage(
  tree: Tree,
  proj: ProjectConfiguration,
  locations: { routes: string },
  options: NormalizedSchema
): void {
  const segments = getSegments(options);
  const routeParams = segments.filter((s) => s.isParam).map((p) => p.path);
  const paramDeclaration =
    routeParams.length > 0
      ? `const { ${routeParams.join(', ')} } = params;`
      : '';

  const answer = segments
    .map((s) => (s.isParam ? `\${${s.path}}` : s.path))
    .join('/');

  switch (options.load) {
    case 'server': {
      const fname = joinPathFragments(
        proj.root,
        locations.routes,
        options.routePath,
        options.language === 'ts' ? '+page.server.ts' : '+page.server.js'
      );

      const ts = `import type { PageServerLoad, RouteParams } from './$types';
      async function demo(params: RouteParams): Promise<string> {
        ${paramDeclaration}
        return Promise.resolve(\`${answer}\`);
      }
      
      export const load = (async({ params }) => {
        return {
          subject: await demo(params)
        };
      }) satisfies PageServerLoad;`;

      const js = `async function demo(params) {
        ${paramDeclaration}
        return Promise.resolve('${options.name}');
      }
      /** @type {import('./$types').PageServerLoad} */
      export async function load({ params }) {
        return {
          subject: await demo(params)
        };
      }
      `;
      tree.write(fname, options.language === 'ts' ? ts : js);
      break;
    }
    case 'shared': {
      const fname = joinPathFragments(
        proj.root,
        locations.routes,
        options.routePath,
        options.language === 'ts' ? '+page.ts' : '+page.js'
      );

      const ts = `import type { PageLoad, RouteParams } from './$types';
      export const load = (async ({ params }) => {
        ${paramDeclaration}
        return {
          subject: \`${answer}\`
        };
      }) satisfies PageLoad;`;
      const js = `/** @type {import('./$types').PageLoad} */
      export function load({ params }) {
        ${paramDeclaration}
        return {
          subject: \`${answer}\`
        };
      }`;

      tree.write(fname, options.language === 'ts' ? ts : js);
      break;
    }
  }
}

function addSveltePage(
  tree: Tree,
  proj: ProjectConfiguration,
  locations: { routes: string },
  options: Required<NormalizedSchema>
): void {
  const fname = joinPathFragments(
    proj.root,
    locations.routes,
    options.routePath,
    '+page.svelte'
  );

  const scripts: Record<
    string,
    { none: string; server: string; shared: string }
  > = {
    js: {
      none: `<script>
        let data = { subject: '${options.name}' };
      </script>`,
      server: `<script>
        /** @type {import('./$types').PageData} */
        export let data;
      </script>`,
      shared: `<script>
        /** @type {import('./$types').PageData} */
        export let data;
      </script>`,
    },
    ts: {
      none: `<script lang="ts">
        let data = { subject: '${options.name}' };
      </script>`,
      server: `<script lang="ts">
        import type { PageData } from './$types';
        export let data: PageData;
      </script>`,
      shared: `<script lang="ts">
        import type { PageData } from './$types';
        export let data: PageData;
      </script>`,
    },
  };

  const script = scripts[options.language][options.load];

  const html = `<article class="container">
	{data.subject} works
</article>
`;

  // TODO add sass devdep in project if selected
  const styles = `<style${options.style === 'scss' ? ' lang="scss"' : ''}>
  .container {
    padding: 1em;
    border: thin solid silver;
  }
  @media screen and (min-width: 576px) { /* landscape phones */ }
  @media screen and (min-width: 768px) { /* tablets */ }
  @media screen and (min-width: 992px) { /* desktops */ }
  @media screen and (min-width: 1200px) { /* large desktops */ }
  @media screen and (min-width: 1400px) { /* larger desktops */ }
</style>`;

  tree.write(fname, [script, html, styles].join('\n\n'));
}
function addPlaywrightPage(
  tree: Tree,
  proj: ProjectConfiguration,
  locations: { tests: string },
  options: NormalizedSchema
): void {
  const fname = joinPathFragments(
    proj.root,
    locations.tests,
    options.directory ?? '',
    `${names(options.name).fileName}.spec.ts`
  );
  // const routePath = joinPathFragments(options.directory ?? '', options.name);

  const routePath = getSegments(options)
    .map((f) => (f.isParam ? `_${f.path}_` : f.path))
    .join('/');

  const content = `import { expect, test } from '@playwright/test';

test('has generated article', async ({ page }) => {
  await page.goto('/${routePath}');
  expect(
    await page.textContent('article.container')
  ).toContain('works');
});
`;
  tree.write(fname, content);
}

export async function routeGenerator(
  tree: Tree,
  schema: Schema
): Promise<void> {
  const notfound = (p: string) => `project '${p}' was not found in workspace`;
  const notSvelte = (p: string) =>
    `project '${p}' is not configured for svelte`;

  const options = normalizeOptions(tree, schema);

  const projects = getProjects(tree);
  const proj = projects.get(options.project);
  if (!proj) {
    throw new Error(notfound(options.project));
  }

  const configContent = getSvelteConfig(tree, proj);

  if (!configContent) {
    throw new Error(notSvelte(options.project));
  }

  const locations = getSvelteFiles(configContent);

  addLoadPage(tree, proj, locations, options);

  addSveltePage(tree, proj, locations, options);

  if (!options.skipTests) {
    addPlaywrightPage(tree, proj, { tests: 'tests' }, options);
  }
  if (!options.skipFormat) {
    await formatFiles(tree);
  }
}

export default routeGenerator;
