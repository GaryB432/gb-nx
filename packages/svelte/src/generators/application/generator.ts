import {
  NX_VERSION,
  addDependenciesToPackageJson,
  ensurePackage,
  formatFiles,
  installPackagesTask,
  joinPathFragments,
  normalizePath,
  updateJson,
  writeJson,
  type GeneratorCallback,
  type Tree,
} from '@nx/devkit';

import { initGenerator } from '@nx/js';

import {
  eslintVersion,
  typescriptESLintVersion,
} from '@nx/eslint/src/utils/versions';
import type {
  NxProjectPackageJsonConfiguration,
  PackageJson,
} from 'nx/src/utils/package-json';
import { updateEslint } from '../../utils/eslint';
import { includes } from '../../utils/globber';
import { isSvelte } from '../../utils/svelte';
import {
  eslintPluginGbVersion,
  eslintPluginSvelteVersion,
  prettierPluginSvelteVersion,
} from '../../utils/versions';
import { normalizeOptions } from './lib/normalize-options';
import { type Config as PrettierConfig } from './lib/prettier';
import {
  type ApplicationGeneratorOptions,
  type NormalizedOptions,
} from './schema';

const PRETTIER_PLUGIN_SVELTE = 'prettier-plugin-svelte';

const nx: NxProjectPackageJsonConfiguration = {
  namedInputs: {
    default: ['{projectRoot}/**/*'],
    production: [
      '!{projectRoot}/.svelte-kit/*',
      '!{projectRoot}/build/*',
      '!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)',
      '!{projectRoot}/tsconfig.spec.json',
    ],
  },
  targets: {
    build: {
      inputs: ['production', '^production'],
      outputs: ['{projectRoot}/build'],
      dependsOn: ['^build'],
    },
  },
};

function addNxConfig(tree: Tree, packageJsonPath: string): void {
  return updateJson<PackageJson>(tree, packageJsonPath, (json) => ({
    ...json,
    nx,
  }));
}

function addScriptsToPackageJson(
  tree: Tree,
  scripts: Record<string, string>,
  packageJsonPath: string
): void {
  updateJson<PackageJson>(tree, packageJsonPath, (json) => {
    json.scripts = json.scripts ?? {};
    for (const script of Object.keys(scripts)) {
      json.scripts[script] = scripts[script];
    }
    return json;
  });
}

function addWorkspaceToPackageJson(
  tree: Tree,
  options: NormalizedOptions,
  packageJsonPath = 'package.json'
): void {
  updateJson<PackageJson>(tree, packageJsonPath, (json) => {
    json.workspaces = json.workspaces ?? [];
    if (Array.isArray(json.workspaces)) {
      if (!includes(options.projectRoot, json.workspaces)) {
        json.workspaces.push(normalizePath(options.projectRoot));
        json.workspaces.sort();
      }
    } else {
      throw new Error('TODO handle');
    }
    return json;
  });
}

function updatePrettier(tree: Tree, options: NormalizedOptions) {
  ensurePackage(PRETTIER_PLUGIN_SVELTE, prettierPluginSvelteVersion);

  const updateConfig = () => {
    const prettierrc = '.prettierrc';
    if (!tree.exists(prettierrc)) {
      writeJson(tree, prettierrc, { singleQuote: true });
    }

    updateJson<PrettierConfig, PrettierConfig>(tree, prettierrc, (json) => {
      json.plugins = json.plugins ?? [];
      if (!json.plugins.includes(PRETTIER_PLUGIN_SVELTE)) {
        json.plugins.push(PRETTIER_PLUGIN_SVELTE);
      }
      json.overrides = json.overrides ?? [];
      if (!json.overrides.some((t) => t.files === '*.svelte')) {
        json.overrides.push({
          files: '*.svelte',
          options: { parser: 'svelte' },
        });
      }
      return json;
    });
  };
  const updateIgnore = () => {
    const fname = '.prettierignore';
    const tbs = ['.svelte-kit', 'build'];
    const newPatterns = tbs.map((p) =>
      joinPathFragments(options.projectRoot, p)
    );

    let content = [
      '# Add files here to ignore them from prettier formatting',
      '/dist',
      '/coverage',
      '/.nx/cache',
    ].join('\n');

    if (tree.exists(fname)) {
      content = tree.read(fname, 'utf-8')!;
    }

    const patterns = content.split('\n');

    const lines = patterns
      .concat(newPatterns.filter((p) => !patterns.includes(p)))
      .concat('');

    tree.write(fname, lines.join('\n'));
  };
  addDependenciesToPackageJson(
    tree,
    {},
    {
      'prettier-plugin-svelte': prettierPluginSvelteVersion,
    }
  );
  updateConfig();
  updateIgnore();
}

export default async function (
  tree: Tree,
  options: ApplicationGeneratorOptions
): Promise<GeneratorCallback> {
  const notSvelte = (p: string) =>
    `project at '${p}' is not configured for svelte`;

  const normalizedOptions = await normalizeOptions(tree, options);
  const config = { root: normalizedOptions.projectRoot };

  if (!isSvelte(tree, config)) {
    throw new Error(notSvelte(normalizedOptions.projectRoot));
  }

  const webPackageJsonPath = joinPathFragments(
    normalizedOptions.projectRoot,
    'package.json'
  );

  await initGenerator(tree, {
    skipFormat: normalizedOptions.skipFormat,
    tsConfigName: 'tsconfig.gb.json',
  });

  addNxConfig(tree, webPackageJsonPath);

  updatePrettier(tree, normalizedOptions);
  addWorkspaceToPackageJson(tree, normalizedOptions, 'package.json');

  normalizedOptions.eslint = false;
  if (normalizedOptions.eslint) {
    // TODO use @nx/eslint (@nx/eslint works with projects. this is not a project yet)
    updateEslint(tree, config);
    addDependenciesToPackageJson(
      tree,
      {},
      {
        '@nx/eslint-plugin': NX_VERSION,
        '@typescript-eslint/eslint-plugin': typescriptESLintVersion,
        '@typescript-eslint/parser': typescriptESLintVersion,
        eslint: eslintVersion,
        'eslint-plugin-gb': eslintPluginGbVersion,
        'eslint-plugin-svelte': eslintPluginSvelteVersion,
      },
      'package.json'
    );
    tree.write(
      joinPathFragments(normalizedOptions.projectRoot, 'tsconfig.base.json'),
      '{ "extends": "./tsconfig.json" }'
    );
    addScriptsToPackageJson(tree, { lint: 'eslint .' }, webPackageJsonPath);
  }

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return async () => {
    installPackagesTask(tree);
  };
}
