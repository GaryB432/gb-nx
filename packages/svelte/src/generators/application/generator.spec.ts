import type { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { type PackageJson } from 'nx/src/utils/package-json';
import { createSvelteKitApp } from '../../utils/svelte';
import generator from './generator';
import type { ApplicationGeneratorOptions } from './schema';

const PRETTIERIGNORE = '.prettierignore';

const installPackagesTask = jest.fn();

jest.mock('@nx/devkit', () => {
  const devkit = { ...jest.requireActual('@nx/devkit') };
  return {
    ...devkit,
    installPackagesTask: jest.fn(),
  };
});

describe('with eslint', () => {
  let appTree: Tree;
  const options: ApplicationGeneratorOptions = { projectPath: 'apps/test', eslint: true };

  const version = '0.0.0-alpha.0';

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace();
    jest.clearAllMocks();
    createSvelteKitApp(appTree, version, {
      name: 'test',
      directory: 'apps',
    });
  });

  it('should handle eslint config', async () => {
    await generator(appTree, options);
    expect(appTree.exists('.eslintrc.json')).toBeTruthy();
  });
  it('should add script', async () => {
    await generator(appTree, options);
    const buff = appTree.read('apps/test/package.json', 'utf-8')!;
    const pj = JSON.parse(buff?.toString()) as unknown as PackageJson;
    expect(pj.scripts!['lint']).toEqual('eslint .');
  });

  it('should add web dev dependencies', async () => {
    await generator(appTree, options);
    const buff = appTree.read('apps/test/package.json', 'utf-8')!;
    const pj = JSON.parse(buff?.toString()) as unknown as PackageJson;
    expect(pj.devDependencies!['@typescript-eslint/parser']).toBeDefined();
  });
  it('should add root dev dependencies', async () => {
    await generator(appTree, options);
    const buff = appTree.read('package.json', 'utf-8')!;
    const pj = JSON.parse(buff?.toString()) as unknown as PackageJson;
    expect(pj.devDependencies!['@nx/eslint-plugin']).not.toBeDefined();
    expect(
      pj.devDependencies!['@typescript-eslint/eslint-plugin']
    ).toBeDefined();
    expect(pj.devDependencies!['@typescript-eslint/parser']).toBeDefined();
    expect(pj.devDependencies!['eslint']).toBeDefined();
    expect(pj.devDependencies!['eslint-plugin-gb']).toBeDefined();
    expect(pj.devDependencies!['eslint-plugin-svelte']).toBeDefined();
  });
});

describe('application generator', () => {
  let appTree: Tree;
  const options: ApplicationGeneratorOptions = {
    projectPath: 'apps/test',
    skipFormat: false,
  };

  const version = '0.0.0-alpha.0';
  const pt: PackageJson = {
    name: 'test',
    version: '0',
    // workspaces: { packages: [] },
    workspaces: ['apps/z', 'apps/a'],
  };

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace();
    appTree.write('package.json', JSON.stringify(pt));
    jest.clearAllMocks();
    createSvelteKitApp(appTree, version, {
      name: 'test',
      directory: 'apps',
    });
  });

  it('should add prettier-plugin-svelte', async () => {
    await generator(appTree, options);
    const p = JSON.parse(appTree.read('package.json')!.toString());
    const v = p.devDependencies['prettier-plugin-svelte'];
    expect(v).toEqual('1.1.1');
  });

  it('should run successfully', async () => {
    await generator(appTree, options);
    // const config = readProjectConfiguration(appTree, 'test');

    const { nx } = JSON.parse(
      appTree.read('apps/test/package.json')!.toString()
    );

    expect(nx).toEqual({
      namedInputs: {
        default: ['{projectRoot}/**/*'],
        production: [
          '!{projectRoot}/.svelte-kit/*',
          '!{projectRoot}/build/*',
          '!{projectRoot}/tests/*',
        ],
      },
      targets: {
        build: {
          inputs: ['production', '^production'],
          outputs: ['{projectRoot}/build'],
          dependsOn: ['^build'],
        },
      },
    });
    // expect(installPackagesTask.mock.calls.length).toBe(2);
  });

  it('should update prettier', async () => {
    appTree.write(PRETTIERIGNORE, '# hi\n\ndist\nstuff\n');
    await generator(appTree, options);
    const p = appTree.read(PRETTIERIGNORE);
    expect(p?.toString()).toMatchInlineSnapshot(`
      "# hi

      dist
      stuff

      apps/test/.svelte-kit
      apps/test/build
      "
    `);
  });

  it('should not update prettier twice', async () => {
    appTree.write(
      PRETTIERIGNORE,
      '# hi\n\ndist\nstuff\n\napps/test/.svelte-kit\napps/test/build\n'
    );
    await generator(appTree, options);
    const p = appTree.read(PRETTIERIGNORE);
    expect(p?.toString()).toMatchInlineSnapshot(`
      "# hi

      dist
      stuff

      apps/test/.svelte-kit
      apps/test/build

      "
    `);
  });

  it('should create prettier', async () => {
    await generator(appTree, options);
    const p = appTree.read(PRETTIERIGNORE);
    expect(p?.toString()).toMatchInlineSnapshot(`
      "# Add files here to ignore them from prettier formatting

      apps/test/.svelte-kit
      apps/test/build
      "
    `);
  });

  it('should add workspace', async () => {
    await generator(appTree, options);
    const p = appTree.read('package.json');
    const s = p ? p.toString() : '';
    const q = JSON.parse(s);
    expect(q.workspaces).toEqual(['apps/a', 'apps/test', 'apps/z']);
  });
});
