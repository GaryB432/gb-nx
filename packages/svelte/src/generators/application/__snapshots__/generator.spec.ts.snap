// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application generator should run successfully 3`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "subject-skapp",
  "namedInputs": {
    "default": [
      "{projectRoot}/**/*",
    ],
    "production": [
      "!{projectRoot}/.svelte-kit/*",
      "!{projectRoot}/build/*",
      "!{projectRoot}/tests/*",
    ],
  },
  "projectType": "application",
  "sourceRoot": "apps/subject-skapp/src",
  "tags": [],
  "targets": {
    "build": {
      "dependsOn": [
        "^build",
      ],
      "inputs": [
        "production",
        "^production",
      ],
      "outputs": [
        "{projectRoot}/build",
      ],
    },
  },
}
`;

exports[`with eslint should handle eslint config 1`] = `
"{
  "root": true,
  "ignorePatterns": [
    "**/*"
  ],
  "plugins": [
    "@nx"
  ],
  "overrides": [
    {
      "files": [
        "*.ts",
        "*.tsx",
        "*.js",
        "*.jsx"
      ],
      "rules": {
        "@nx/enforce-module-boundaries": [
          "error",
          {
            "enforceBuildableLibDependency": true,
            "allow": [],
            "depConstraints": [
              {
                "sourceTag": "*",
                "onlyDependOnLibsWithTags": [
                  "*"
                ]
              }
            ]
          }
        ]
      }
    },
    {
      "files": [
        "*.ts",
        "*.tsx"
      ],
      "extends": [
        "plugin:@nx/typescript"
      ],
      "rules": {}
    },
    {
      "files": [
        "*.js",
        "*.jsx"
      ],
      "extends": [
        "plugin:@nx/javascript"
      ],
      "rules": {}
    }
  ]
}
"
`;

exports[`with eslint should handle project eslint config 1`] = `
"{
  "extends": [
    "../../.eslintrc.json",
    "plugin:@typescript-eslint/recommended",
    "plugin:svelte/recommended"
  ],
  "ignorePatterns": [
    "!**/*",
    "node_modules/*",
    ".svelte-kit/*"
  ],
  "overrides": [
    {
      "files": [
        "*.ts",
        "*.tsx",
        "*.js",
        "*.jsx"
      ],
      "rules": {}
    },
    {
      "files": [
        "*.ts",
        "*.tsx"
      ],
      "rules": {}
    },
    {
      "files": [
        "*.js",
        "*.jsx"
      ],
      "rules": {}
    },
    {
      "files": [
        "*.svelte"
      ],
      "parser": "svelte-eslint-parser",
      "parserOptions": {
        "parser": "@typescript-eslint/parser"
      },
      "rules": {}
    }
  ]
}
"
`;
